### PR TITLE
Add add_storage() method to UnitSentry

### LINKED ISSUE #112: Add Support for Juju Storage
This is needed for testing charms that depend on the juju-storage feature, these charms come with
-  [name]-storage-attached
-  [name]-storage-detaching
hooks for every used storage relation, which need testing.

A charm can be deployed with different storage provider (e.g. loop or ebs for block)
`$ juju deploy <charm> --storage <name>=<pool>,<size>,count`

This could be expressed as an extension of  Deployment.add(), e.g.
`.add( 'servicename', storage={servicestorage:{'pool': 'ebs', 'size': 1024, 'count': 3}})`

Juju-Storage also provides the possibility to add storage to already deployed units 
`$ juju storage add <unit> <name>=<pool>,<size>,count`

I'm not sure how the adding (removing) of storage should be expressed in terms of amulet.

Trying to add this support myself, I stopped when I realize that amulet makes use of juju-deployer. juju-deployer does not provide support for juju-storage (not sure if it is expressible in the bundle-format, was not able to create a bundle using juju-gui export on a deployment with juju-storage)

### PR DESCRIPTION
**Pull request by tvansteenburgh**
Fixes #112.

Adds an add_storage() method the UnitSentry, e.g.:

```python
unit = d.sentry['wordpress'][0]
unit.add_storage('mystorage', pool='ebs', count=3, size=1024)
```

There are no 'remove storage' commands in Juju.

### CODE REVIEW

**Comment by AdamIsrael** [comment]
Would it make sense to add methods for list-storage as well?

Otherwise, this LGTM.

**Comment by marcoceppi** [comment]
@Yrrsinn Will this be sufficient?

**Comment by Yrrsinn** [comment]
I am not sure. This patch covers the case of adding storage to a deployed charm, that is true. But I see two points here (both related to storage-pool, but different aspects). 

First, the approach forces the charm-test-author to explicitly name a storage provider, e.g. ebs. Shouldn't a charm-test be agnostic about the execution environment? More explicit: Would a test that has `pool='ebs'` hard-wired pass on a juju/maas test-environment? Is this a use-case?

Second, a charm can express that it needs at least one volume when deployed (non-optional storage). The constraints about the storage pool to use, should be expressible. Compare to `https://lists.ubuntu.com/archives/juju/2016-April/006987.html`.

The PR wraps around parts of the command-line interface. Imho a better approach would be to use the support for storage in bundle deploy (added in 2.0 afaik), since this is the general approach of amulet.

**Comment by tvansteenburgh** [comment]
@Yrrsinn, good feedback, thanks.

Regarding your first point: For add_storage(), if `pool=None` then the default pool for the provider will be used, which may be sufficient for test cases. If you need to create non-default pools for your tests we'll need to add a

```python
Deployment.create_storage_pool(name='mypool', provider_type='ebs', **config)
```

method, which *will* need to know the provider type at run time. One way to achieve this is to support AMULET_STORAGE_PROVIDER and AMULET_STORAGE_CONFIG environment variables that could be passed in at runtime. Thoughts?

Regarding your second point: The reason we used CLI commands is because there is no native bundle support for juju-1.25, and juju-deployer will possibly never support storage, and we want amulet to work for both juju1 and juju2. Are there use cases that are impossible with the current implementation?

**Comment by AdamIsrael** [comment]
@Yrrsinn, what kind of problems did you run into when setting the root-disk machine constraint large enough for the 10GB loop device? I'm running bundletester against the quobyte bundle now, with the increased constraints, and I'd like to see if I encounter the same result you did.

**Comment by Yrrsinn** [comment]
@AdamIsrael it simply did not work? Honestly, I tried this in January 2016, and filed [this issue #112](https://github.com/juju/amulet/issues/112) a few days later. Since we have the mid of June by now, I do not know what was the precise error message back then.

Maybe the problem is fixed meanwhile by one of the Juju 1.25.x bug-fix releases, or it was related to the problems with the Juju Vagrant Box, Canonical advertises on [jujucharms.com/docs/stable](https://jujucharms.com/docs/stable/config-vagrant).

Actually I tried to upgrade from the trusty to the wily Vagrant Box in January, to fix 'strange behaviour' (a.k.a. Juju did not work as supposed). Unfortunately the wily Vagrant Box was broken, and and the bug I filed for that issue on February 2th is still in the state 'new'  on June 15th: [Juju vagrant box failed to start (wily)](https://bugs.launchpad.net/juju-vagrant-images/+bug/1540471). Meanwhile I use a Docker based set-up.

Just this week, we learned that [running tests with bundletester on charms with storage can cause the bundletester to crash.](https://bugs.launchpad.net/juju-core/+bug/1592366)

tl;dr
I do not have the time to debug all your software. I am supposed to introduce 'bugs' in our own software.

**Comment by AdamIsrael** [comment]
@Yrrsinn I'll follow up on those bugs you listed. Thanks for filing them, and sorry they haven't been addressed sooner.

Right now, I'm testing with juju 2.0 beta7/8, using a vagrant box running Xenial. We still recommend the Trusty vagrant juju box because the others are not as well-tested. That's something we're working to address in the near future.

The test errors I'm seeing, running bundletester against the bundle, are timeouts waiting for the environment to stand up. I'd really like to see those timeouts bumped up per our email discussion, because I think they'll resolve many of the problems I've seen, at least as far as Juju 2 is concerned. I'll verify that today, and if everything does work as expected, I'll raise the issue of backporting the fixes in Juju 2 to the 1.25 series.
